packages/@aws-cdk/aws-ec2/README.md

```diff
@@ -551,17 +556,73 @@ new VpcEndpointService(this, 'EndpointService', {
 });
 ```
 
-## Bastion Hosts
+## Instances
+
+You can use the `Instance` class to start up a single EC2 instance. For production setups, we recommend
+you use an `AutoScalingGroup` from the `aws-autoscaling` module instead, as AutoScalingGroups will take
+care of restarting your instance if it ever fails.
+
+### Configuring Instances using CloudFormation Init (cfn-init)
+
+CloudFormation Init allows you to configure your instances by writing files to them, installing software
+packages, starting services and running arbitrary commands. By default, if any of the instance setup
+commands throw an error, the deployment will fail and roll back to the previously known good state.
+The following documentation also applies to `AutoScalingGroup`s.
+
+For the full set of capabilities of this system, see the documentation for
+[`AWS::CloudFormation::Init`](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-init.html).
+Here is an example of applying some configuration to an instance:
+
+```ts
+new ec2.Instance(this, 'Instance', {
+  init: ec2.CloudFormationInit.fromElements(
+    ec2.InitCommand.shellCommand('/bin/true'),
+  ),
+  initOptions: {
+    // Optional, which configsets to activate (['default'] by default)
+    configSets: ['default'],
+
+    // Optional, how long the installation is expected to take (5 minutes by default)
+    timeout: Duration.minutes(30),
+  },
+
+  // Optional but recommended: starts a new instance with the new config
+  // if the config changes.
+  userDataCausesReplacement: true,
+});
+```
+
+You can have services restarted after the init process has made changes to the system.
+To do that, instantiate an `InitServiceRestartHandle` and pass it to the config elements
+that need to trigger the restart and the service itself. For example, the following
+config installs nginx through a custom script, and then
+restarts nginx so that it picks up the new config and files:
+
+```ts
+const handle = new ec2.InitServiceRestartHandle();
```

Comment:
@rix0rrr  and I debated a few different ways to solve this, and this was the best we came up with. I'm open to suggestions otherwise. We either need (a) an object that relates an InitService to InitFile, InitCommand, etc. objects or (b) Ask the user to re-specify all of the strings used for the InitFile, InitCommand, etc. objects in the InitService object. So I believe our original take on this didn't have the restart handle, but the interface looked like this:

```ts
ec2.InitService.enable({
  commands: ['/usr/bin/custom-nginx-install.sh'],
  files: ['...'],
});
```

Where the user is forced to re-specify all of the "keys" for any triggering objects. I think the current implementation is cleaner than that, but open to debate. Also see my back & forth with @nija-at  on the same topic.

Reply:
This is somewhat analogous to how we use "artifacts" in CodePipeline to establish references between different actions. Roughly:

```ts
const sourceArtifact = new codepipeline.Artifact();

new Pipeline(..., {
  actions: [
    new GitHubSourceAction({ output: sourceArtifact }),
    new BuildAction({ input: sourceArtifact }),
  ]
});
```

The alternative there was:

```ts
const github = new GitHubSourceAction({ ... });

new Pipeline(..., {
  actions: [
    github,
    new BuildAction({ input: github.output }),
  ]
});
```

Which from a type safety standpoint is better, but from a code readability standpoint is worse (imo).

We compensate for the type safety by using runtime checks.